timemachine/fe/free_energy.py

```diff
@@ -835,23 +847,48 @@ def assert_ensembles_compatible(state_a: InitialState, state_b: InitialState):
     warn(f"Setting numpy global random state using seed {md_params.seed}")
     np.random.seed(md_params.seed)
 
-    bps = initial_states[0].potentials  # TODO: assert initial states have compatible potentials?
-    sp = SummedPotential([bp.potential for bp in bps], [bp.params for bp in bps])
-    ubp = sp.to_gpu(np.float32)
+    def get_potential_and_context(initial_state: InitialState) -> Tuple[GpuImplWrapper, Context]:
+        # Set up overall potential
+        potentials = [bp.potential for bp in initial_state.potentials]
+        params = [bp.params for bp in initial_state.potentials]
+        potential = SummedPotential(potentials, params).to_gpu(np.float32)
+
+        # Set up context for MD using overall potential
+        bound_impl = potential.bind_params_list(params).bound_impl
+        bound_impls = [bound_impl]
+        intg_impl = initial_state.integrator.impl()
+        baro_impl = initial_state.barostat.impl(bound_impls) if initial_state.barostat else None
+        context = Context(
+            initial_state.x0,
+            initial_state.v0,
+            initial_state.box0,
+            intg_impl,
+            bound_impls,
+            baro_impl,
+        )
+
+        return potential, context
 
-    def compute_log_q_matrix(xvbs: List[CoordsVelBox], params: List[NDArray]):
+    # Set up overall potential and context using the first state.
+    # NOTE: this assumes that states differ only in their parameters, but we do not check this!
```

Comment:
tbh I think this is a reasonable approach, I don't see us doing TREX/PREX where the temperature/pressure will change in the standard way since the energies are size extensive.